include/inviwo/core/util/indirectiterator.h

```diff
@@ -0,0 +1,225 @@
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+#ifndef IVW_INDIRECTITERATOR_H
+#define IVW_INDIRECTITERATOR_H
+
+#include <iterator>
+#include <type_traits>
+
+namespace inviwo {
+
+namespace util {
+
+namespace detail_indirect {
+template <typename Tag, typename Iter>
+struct require : std::enable_if<std::is_base_of<Tag, typename Iter::iterator_category>::value> {};
+template <typename Tag, typename Iter>
+using require_t = typename require<Tag, Iter>::type;
+
+template <typename Iterator>
+struct is_const_iterator {
+    typedef typename std::iterator_traits<Iterator>::pointer pointer;
+    static constexpr bool value = std::is_const<typename std::remove_pointer<pointer>::type>::value;
+};
+
+template <typename T>
+struct add_const_to_reference {
+    using type = const T;
+};
+
+template <typename T>
+struct add_const_to_reference<T&> {
+    using type = const T&;
+};
+
+template <typename T>
+struct add_const_to_reference<T&&> {
+    using type = const T&&;
+};
+
+template <typename T>
+using add_const_to_reference_t = typename add_const_to_reference<T>::type;
+
+template <bool asConst>
+struct asPointer {};
+
+template <>
+struct asPointer<true> {
+    template <typename T>
+    static constexpr const T* get(const T* p) {
+        return p;
+    }
+    template <typename T, typename D>
+    static constexpr const T* get(const std::unique_ptr<T, D>& p) {
+        return p.get();
+    }
+    template <typename T>
+    static constexpr const T* get(const std::shared_ptr<T>& p) {
+        return p.get();
+    }
+};
+template <>
+struct asPointer<false> {
+    template <typename T>
+    static constexpr T* get(const T* p) {
+        return p;
+    }
+    template <typename T, typename D>
+    static constexpr T* get(const std::unique_ptr<T, D>& p) {
+        return p.get();
+    }
+    template <typename T>
+    static constexpr T* get(const std::shared_ptr<T>& p) {
+        return p.get();
+    }
+};
+
+}  // namespace detail_indirect
+
+template <typename Iter, bool PropagateConst = true>
+struct IndirectIterator {
```

Comment:
Could add PR comment as documentation here to explain what Indirect refers to.

Reply:
Will do